app/models/user.rb

```diff
@@ -18,6 +19,38 @@ class User < ApplicationRecord
                          wrong_length: 'must be between 3 and 20 characters'
                        }
 
+  class << self
+    def from_omniauth(auth)
+      username = auth.info.email.split('@').first
+
+      find_or_create_by(
+        provider: auth.provider,
+        uid: auth.uid,
+        params: new_user_params(auth, username)
+      )
+    end
+
+    private
+
+    def can_use_username?(username, email)
+      !find_by(username:, email:).blank?
+    end
+
+    def new_user_params(auth, username)
+      {
+        username: can_use_username?(username, auth.info.email) ? username : "johndoe#{rand(1000)}",
+        email: auth.info.email,
+        password: Devise.friendly_token[0, 20],
+        full_name: auth.info.name,
+        avatar_url: auth.info.image
+      }
+    end
+
+    def find_or_create_by(provider:, uid:, params:)
+      where(provider:, uid:).first_or_create(params)
+    end
```

Comment:
It looks like this method is overriding the `User.find_or_create_by` method that comes from ActiveRecord, I don't think it's a good idea. Maybe you could rename this method? Or even better, I think you can extract all this functionality in a separate class, like `OmniauthAuthentication` or something like that, because all the logic you are adding here is very specific for this feature what do you think?